File: src/env.py
import gymnasium as gym 
import numpy as np
import matplotlib.pyplot as plt
import random
import itertools

class GridEnvDeform(gym.Env):
    def __init__(self, maze, l0,h0,l1,h1):

        self.original_maze = maze
        self.original_maze_shape = maze.shape
 
        self.maze = maze
        self.maze_shape = maze.shape

        # list of possible actions
        self.actions = [0,1,2,3]
        # list of possible orientations
        self.orientations = [0,1,2,3]
        # list of possible observations
        self.obs = list(itertools.product([0,1], repeat=5))
        # list of possible deformations
        self.deformations = [(i,j) for i in range(l0,h0) for j in range(l1,h1)]
        
        # space in which every maze lives (is a 2d matrix)
        self.max_shape = self.original_maze.shape * np.array([h1-1,h0-1]) + np.array([2,2])
        # list of states
        self.states = [((x,y,phi),(i,j)) for x in range(1,self.max_shape[0]-1) for y in range(1,self.max_shape[1]-1) for phi in range(4) for i in range(l0,h0) for j in range(l1,h1)] 
                

        self.goal_pos = self.original_maze.shape - np.array([2,2])
        
        self.reset()
        
    def T(self, s, a, s_):
        """ s = [(x, y, phi), thetas]
            T(s,a,s_) transition probability from s to s_ given action a 
            assuming is deterministic and always possible to move in the direction of the action (except at the boundaries)
        """
        # if the maze deformation is different the transition is impossible       
        if s[1] != s_[1]:
            return 0.
                
        return 1. if np.all(self.step(a,s)[0] == s_) else 0

    def O(self, s, a, o):
        # observation probability

        return 1 if np.all(self.get_observation(s) == o) else 0
    
    def R(self, s, a, s_=None):
        
        """R(s,a,s_) reward for getting to s_ (independent from previous state and action taken)"""
            
        return self.step(a,s)[1]            
    
    def step(self, a, s=None, execute=False):

        """take action a from state s (if given) or from actual state of the maze """
        
        if s is not None:
            self.set_deformed_maze(s[1])
            x, y, phi = s[0][:3]
            x_, y_, phi_ = x, y, phi
        else:
            x, y = self.agent_pos
            phi = self.agent_orientation
            x_, y_, phi_ = x, y, phi

        actual_action = (a + phi) % 4
        
        if actual_action == 0:  # Move up
            new_pos = [x - 1, y]
        elif actual_action == 2:  # Move down
            new_pos = [x + 1, y]
        elif actual_action == 3:  # Move left
            new_pos = [x, y - 1]
        elif actual_action == 1:  # Move right
            new_pos = [x, y + 1]
        else:
            raise ValueError("Invalid Action")
        
        # Check if the new position is valid (inside the maze and not a wall)
        if 0 < new_pos[0] < self.max_shape[0]-1 and 0 < new_pos[1] < self.max_shape[1]-1:
            x_, y_ = new_pos
            if execute:
                self.agent_pos = new_pos

        phi_ = (phi + a) % 4
        
        if execute:
            self.agent_orientation = phi_
        
        terminated = np.all((x_,y_) == self.goal_pos)

        if np.all((x_,y_) == self.goal_pos):
            # if the agent is in the goal position
            reward =  10            
        elif np.all((x_,y_) == (x,y)):
            # if the agent has not moved (only at the boundary of the maze)
            reward =  -100/(self.max_shape[0]*self.max_shape[1])
        elif self.maze[x_, y_] == 1:
            # if the agent has entered a wall
            reward =  -100/(self.max_shape[0]*self.max_shape[1])
        elif self.maze[x_, y_] == 0:
            # if the agent has moved to a free cell
            reward =  -1/(self.max_shape[0]*self.max_shape[1])

        info = {}
        truncated = False 

        s_ = ((x_, y_, phi_), self.theta)
        
        if execute:
            self.timestep += 1
        
        return s_, reward, terminated, truncated, info, 
    
    def get_observation(self, s):

        self.set_deformed_maze(s[1])

        agent_pos = s[0][:2]
        agent_orientation = s[0][2]
        ind = [agent_pos + a for a in [np.array([0,-1]),
                                            np.array([-1,-1]),
                                            np.array([-1,0]),
                                            np.array([-1,+1]),
                                            np.array([0,+1]),
                                            np.array([+1,+1]),
                                            np.array([+1,0]),
                                            np.array([+1,-1])]]

        agent_obs = np.array([self.maze[tuple(ind[i%8])] 
                                                for i in range(2*agent_orientation, 2*agent_orientation+5)])

        
        return agent_obs
    
    def set_deformed_maze(self,thetas: tuple):
        self.theta = thetas
        self.maze = self.stretch_maze(thetas)
        # self.goal_pos = self.maze.shape - np.array([thetas[1],thetas[0]])
        self.goal_pos = self.original_maze.shape * np.array([thetas[1],thetas[0]])

        canva1 = np.ones(self.max_shape, dtype=int)  # Start with walls
        # Place the original maze in the canvas
        canva1[1:self.maze.shape[0] + 1, 1:self.maze.shape[1] + 1] = self.maze

        self.maze = canva1
   
    def stretch_maze(self, thetas):
        scale_x, scale_y = thetas
        maze = self.original_maze

        original_height, original_width = maze.shape
        # Calculate new dimensions
        new_height = original_height * scale_y
        new_width = original_width * scale_x
        
        # Create a new maze with stretched dimensions
        stretched_maze = np.ones((new_height, new_width), dtype=int)

        # Fill the new maze with values from the original maze
        for i in range(original_height):
            for j in range(original_width):
                if maze[i, j] == 0:  # Path cell
                    # Fill the corresponding region in the stretched maze
                    stretched_maze[i*scale_y:(i+1)*scale_y, j*scale_x:(j+1)*scale_x] = 0

        return stretched_maze
    
    def set_state(self, s):
        theta0, theta1 = s[1][0], s[1][1]
        self.theta = (theta0, theta1)
        self.agent_pos = np.array(s[0][:2]) 
        self.agent_orientation = s[0][2]
        self.set_deformed_maze(s[1])
    
    def get_state(self):
        return (self.agent_pos[0],self.agent_pos[1], self.agent_orientation), self.theta
    
    def render(self, s=None, s_prime=None):
        """
        belief will be always distributed as follows 
        there exist a tuple (x_0,y_0,phi_0) such that
        for the two possible deformations theta0 theta1

        b((x_0,y_0,phi_0), theta_0) = b((x_0,y_0,phi_0), theta_1) = 0.5
        and 
        b = 0 everywere else
        
        rendering will show the same position in the two different possibile maze deformations
        along with the probability of each status
        """
        if s is not None:
            self.set_state(s)
        
        maze_render = np.copy(self.maze)
        maze_render[tuple(self.agent_pos)] = 2  # Show agent position
        maze_render[tuple(self.goal_pos)] = 4  # Show goal position
        plt.imshow(maze_render, cmap='binary', origin='upper')
        plt.show()

    def reset(self, seed=42):
        randomdeformation = random.choice(self.deformations)
        self.agent_pos = [np.random.randint(1, self.max_shape[0]-1), np.random.randint(1, self.max_shape[1]-1)]
        self.agent_orientation = random.choice(self.orientations)
        self.set_deformed_maze(randomdeformation)
        self.goal_pos = self.original_maze.shape * np.array([randomdeformation[1],randomdeformation[0]])
        self.theta = randomdeformation
        self.timestep = 0
        return (self.agent_pos[0],self.agent_pos[1], self.agent_orientation), (self.maze.shape[0], self.maze.shape[1])
        
    def is_done(self):
        return np.all(self.agent_pos == self.goal_pos)
    
    def is_new(self):
        return self.timestep == 0
    
def create_maze(dim):
    maze = np.ones((dim*2+1, dim*2+1))
    x, y = (0, 0)
    maze[2*x+1, 2*y+1] = 0
    stack = [(x, y)]
    
    while len(stack) > 0:
        x, y = stack[-1]
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        random.shuffle(directions)

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if nx >= 0 and ny >= 0 and nx < dim and ny < dim and maze[2*nx+1, 2*ny+1] == 1:
                maze[2*nx+1, 2*ny+1] = 0
                maze[2*x+1+dx, 2*y+1+dy] = 0
                stack.append((nx, ny))
                break
        else:
            stack.pop()

    # Create entrance and exit
    # maze[1, 0] = 0
    # maze[-2, -1] = 0

    return maze[1:-1, 1:-1]

File: src/environment/cpp_env/env.cpp
#include <iostream>
#include <Eigen/Dense>

#include <iostream>
#include <vector>
#include <fstream>
#include <random>
#include <cmath>
#include <Eigen/Dense>
#include <algorithm>
#include <tuple>
#include <state.cpp>


// g++ env.cpp -o env -I /usr/local/include/eigen3

class GridEnvDeform {
public:
    GridEnvDeform(const Eigen::MatrixXi& maze, int l0, int h0, int l1, int h1);

    std::tuple<std::tuple<int, int, int>, std::tuple<int, int>> reset();
    std::tuple<std::tuple<int, int, int>, int, bool, bool, std::string> step(int action);
    void render();
    bool is_done();
    std::vector<int> get_observation();
    std::tuple<std::tuple<int,int>> get_goal_pos();
    // Helper functions
    Eigen::MatrixXi stretch_maze(const std::tuple<int, int>& thetas);
    void set_deformed_maze(const std::tuple<int, int>& thetas);
    void set_state(const std::tuple<int, int, int>& state, const std::tuple<int, int>& thetas);
    std::tuple<int, int> agent_pos;
    int agent_orientation;
    int max_shape[2];

    
private:
    Eigen::MatrixXi original_maze;
    Eigen::MatrixXi maze;
    std::vector<int> actions = {0, 1, 2, 3}; // up, right, down, left
    std::vector<int> orientations = {0, 1, 2, 3};
    std::vector<std::tuple<int, int>> deformations;
    Eigen::Vector2i goal_pos;
    int l0, h0, l1, h1;
    int timestep;

};


GridEnvDeform::GridEnvDeform(const Eigen::MatrixXi& maze, int l0, int h0, int l1, int h1)
    : original_maze(maze), maze(maze), l0(l0), h0(h0), l1(l1), h1(h1), timestep(0) {

    // Define the list of possible deformations
    for (int i = l0; i < h0; ++i) {
        for (int j = l1; j < h1; ++j) {
            deformations.emplace_back(i, j);
        }
    }

    goal_pos = Eigen::Vector2i(maze.rows() - 2, maze.cols() - 2);
    max_shape[0] = original_maze.rows() * (h1-1) + 2;
    max_shape[1] = original_maze.cols() * (h0-1) + 2 ;

}

Eigen::MatrixXi GridEnvDeform::stretch_maze(const std::tuple<int, int>& thetas) {
    int scale_x = std::get<0>(thetas);
    int scale_y = std::get<1>(thetas);

    int new_height = original_maze.rows() * scale_y;
    int new_width = original_maze.cols() * scale_x;
    Eigen::MatrixXi stretched_maze = Eigen::MatrixXi::Ones(new_height, new_width);

    for (int i = 0; i < original_maze.rows(); ++i) {
        for (int j = 0; j < original_maze.cols(); ++j) {
            if (original_maze(i, j) == 0) {
                stretched_maze.block(i * scale_y, j * scale_x, scale_y, scale_x).setZero();
            }
        }
    }

    return stretched_maze;
}

void GridEnvDeform::set_deformed_maze(const std::tuple<int, int>& thetas) {
    maze = stretch_maze(thetas);
    goal_pos = Eigen::Vector2i(original_maze.rows() * std::get<1>(thetas)-2, original_maze.cols() * std::get<0>(thetas)-2);
    std::cout << "Goal position: " << goal_pos << std::endl;
}

void GridEnvDeform::set_state(const std::tuple<int, int, int>& state, const std::tuple<int, int>& thetas) {
    agent_pos = std::make_tuple(std::get<0>(state), std::get<1>(state));
    agent_orientation = std::get<2>(state);
    set_deformed_maze(thetas);
}

std::tuple<std::tuple<int, int, int>, std::tuple<int, int>> GridEnvDeform::reset() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis_x(1, maze.rows() - 2);
    std::uniform_int_distribution<> dis_y(1, maze.cols() - 2);
    std::uniform_int_distribution<> dis_orient(0, 3);
    std::uniform_int_distribution<> dis_deform(0, deformations.size() - 1);

    int random_idx = dis_deform(gen);
    auto random_deformation = deformations[random_idx];
    agent_pos = std::make_tuple(dis_x(gen), dis_y(gen));
    // goal_pos = Eigen::Vector2i(maze.rows() - 2, maze.cols() - 2);
    agent_orientation = dis_orient(gen);

    set_deformed_maze(random_deformation);
    timestep = 0;

    return {std::make_tuple(std::get<0>(agent_pos), std::get<1>(agent_pos), agent_orientation), random_deformation};
}

std::tuple<std::tuple<int, int, int>, int, bool, bool, std::string> GridEnvDeform::step(int action) {
    // Action mapping based on orientation
    int actual_action = (action + agent_orientation) % 4;

    int x = std::get<0>(agent_pos);
    int y = std::get<1>(agent_pos);
    int new_x = x, new_y = y;

    // Move based on action
    if (actual_action == 0) {  // Move up
        new_x -= 1;
    } else if (actual_action == 1) {  // Move right
        new_y += 1;
    } else if (actual_action == 2) {  // Move down
        new_x += 1;
    } else if (actual_action == 3) {  // Move left
        new_y -= 1;
    }

    // Ensure new position is within bounds and not a wall
    bool terminated = false;
    int reward = -1;
    if (new_x > 0 && new_x < maze.rows() - 1 && new_y > 0 && new_y < maze.cols() - 1) {
        agent_pos = std::make_tuple(new_x, new_y);
        reward = -0.5;
        if (new_x == goal_pos(0) && new_y == goal_pos(1)) {
            terminated = true;
            reward = 1;
        }
    } else {
        reward = -2; // penalty for hitting a borderwall
    }

    // Update orientation
    agent_orientation = (agent_orientation + action) % 4;

    return {std::make_tuple(std::get<0>(agent_pos), std::get<1>(agent_pos), agent_orientation), reward, terminated, false, ""};
}

void GridEnvDeform::render() {
    std::cout << "Maze:" << std::endl;
    for (int i = 0; i < maze.rows(); ++i) {
        for (int j = 0; j < maze.cols(); ++j) {
            if (i == std::get<0>(agent_pos) && j == std::get<1>(agent_pos)) {
                if(agent_orientation == 1){
                    std::cout << "> ";
                }else if(agent_orientation == 2){
                    std::cout << "v ";
                }else if(agent_orientation == 3){
                    std::cout << "< ";
                }else{
                    std::cout << "^ ";
                }
            } else if (i == goal_pos(0) && j == goal_pos(1)) {
                std::cout << "G ";
            } else if (maze(i, j) == 0) {
                std::cout << ". ";
            } else {
                std::cout << "# ";
            }
        }
        std::cout << std::endl;
    }

    std::cout << "Agent position: (" << std::get<0>(agent_pos) << ", " << std::get<1>(agent_pos) << ")" << std::endl;
    std::cout << "Agent orientation: " << agent_orientation << std::endl;
    std::cout << "Goal position: (" << goal_pos(0) << ", " << goal_pos(1) << ")" << std::endl;
}


// void test_get_observation(GridEnvDeform& env) {
//     std::cout << "\nTesting get_observation() function..." << std::endl;
//     auto observation = env.get_observation();

//     std::cout << "Observation around the agent: ";
//     for (int obs : observation) {
//         std::cout << obs << " ";
//     }
//     std::cout << std::endl;
//     std::cout << "Get observation test passed." << std::endl;
// }


int main() {
    // Initialize a simple maze for testing
    Eigen::MatrixXi maze(5, 5);
    maze << 1, 1, 1, 1, 1,
            1, 0, 1, 0, 1,
            1, 0, 1, 0, 1,
            1, 0, 0, 0, 1,
            1, 1, 1, 1, 1;

    int l0 = 1, h0 = 3;
    int l1 = 1, h1 = 2;

    // Create GridEnvDeform object
    GridEnvDeform env(maze, l0, h0, l1, h1);
    env.reset();
    env.render();



    // define states 

    int max_shape_x = env.max_shape[0];  // You'll need to replace with actual values
    int max_shape_y = env.max_shape[1];


    // Generate states
    std::vector<State> states;
    for (int x = 1; x < max_shape_x - 1; x++) {
        for (int y = 1; y < max_shape_y - 1; y++) {
            for (int phi = 0; phi < 4; phi++) {
                for (int i = l0; i < h0; i++) {
                    for (int j = l1; j < h1; j++) {
                        State state;
                        state.pos = std::make_tuple(x, y, phi);
                        state.theta = std::make_pair(i, j);
                        states.push_back(state);
                    }
                }
            }
        }
    }

    std::cout << "States generated: " << states.size() << std::endl;
    // show first 10 states
    for (int i = 0; i < 10; i++) {
        std::cout << "State " << i << ": (" << std::get<0>(states[i].pos) << ", " << std::get<1>(states[i].pos) << ", " << std::get<2>(states[i].pos) << "), (" << states[i].theta.first << ", " << states[i].theta.second << ")" << std::endl;
    }


    // Generate actions
    std::vector<int> actions = {0, 1, 2, 3};


    return 0;
}
